pkg/api/routes.go

```diff
@@ -6,46 +6,44 @@ import (
 	"github.com/gin-gonic/gin"
 )
 
-// GET /api/v1/metrics/dojo-worker-count
 func LoginRoutes(router *gin.Engine) {
 	docs.SwaggerInfo.BasePath = "/api/v1"
 	apiV1 := router.Group("/api/v1")
 	{
 		worker := apiV1.Group("/worker")
 		{
+			worker.Use(GenerousRateLimiter())
 			worker.POST("/login/auth", WorkerLoginMiddleware(), WorkerLoginController)
 			worker.POST("/partner", WorkerAuthMiddleware(), WorkerPartnerCreateController)
 			worker.PUT("/partner/disable", WorkerAuthMiddleware(), DisableMinerByWorkerController)
 			worker.GET("/partner/list", WorkerAuthMiddleware(), GetWorkerPartnerListController)
 		}
-		apiV1.GET("/auth/:address", GenerateNonceController)
-		apiV1.PUT("/partner/edit", WorkerAuthMiddleware(), UpdateWorkerPartnerController)
+		apiV1.GET("/auth/:address", GenerousRateLimiter(), GenerateNonceController)
+		apiV1.PUT("/partner/edit", GenerousRateLimiter(), WorkerAuthMiddleware(), UpdateWorkerPartnerController)
 		tasks := apiV1.Group("/tasks")
 		{
-			tasks.POST("/create-tasks", MinerAuthMiddleware(), CreateTasksController)
-			tasks.PUT("/submit-result/:task-id", WorkerAuthMiddleware(), SubmitTaskResultController)
-			tasks.GET("/task-result/:task-id", GetTaskResultsController)
-			tasks.GET("/:task-id", GetTaskByIdController)
-			tasks.GET("/next-task/:task-id", WorkerAuthMiddleware(), GetNextInProgressTaskController)
-			tasks.GET("/", WorkerAuthMiddleware(), GetTasksByPageController)
+			tasks.PUT("/submit-result/:task-id", WriteTaskRateLimiter(), WorkerAuthMiddleware(), SubmitTaskResultController)
+			tasks.POST("/create-tasks", InMetagraphOnly(), WriteTaskRateLimiter(), MinerAuthMiddleware(), CreateTasksController)
+			tasks.GET("/task-result/:task-id", ReadTaskRateLimiter(), GetTaskResultsController)
+			tasks.GET("/:task-id", ReadTaskRateLimiter(), GetTaskByIdController)
+			tasks.GET("/next-task/:task-id", ReadTaskRateLimiter(), GetNextInProgressTaskController)
+			tasks.GET("/", ReadTaskRateLimiter(), WorkerAuthMiddleware(), GetTasksByPageController)
 		}
 
 		miner := apiV1.Group("/miner")
 		{
-			// miner.POST("/login/auth", MinerLoginMiddleware(), MinerLoginController)
-			// miner.GET("/info/:hotkey", MinerAuthMiddleware(), MinerInfoController)
-			// miner.POST("/miner-application", MinerVerificationMiddleware(), MinerApplicationController)
-			// miner.PUT("/partner/disable", MinerAuthMiddleware(), DisableWorkerByMinerController)
-			miner.POST("/session/auth", GenerateCookieAuth)
+			miner.POST("/session/auth", GenerousRateLimiter(), GenerateCookieAuth)
 
 			apiKeyGroup := miner.Group("/api-key")
+			apiKeyGroup.Use(GenerousRateLimiter())
```

Comment:
btw, these GenerousRateLimiter will share resources across the route. For those routes which won't get trigger very much, it will be fine though.